executor/src/witgen/data_structures/finalizable_data.rs

```diff
@@ -4,47 +4,118 @@ use std::{
 };
 
 use bit_vec::BitVec;
-use powdr_ast::analyzed::PolyID;
+use itertools::Itertools;
+use powdr_ast::analyzed::{PolyID, PolynomialType};
 use powdr_number::FieldElement;
 
 use crate::witgen::rows::Row;
 
-/// A row entry in [FinalizableData].
+/// Sequence of rows of field elements, stored in a compact form.
 #[derive(Clone)]
-enum Entry<T: FieldElement> {
-    /// The row is still in progress, and range constraints are still available.
-    InProgress(Row<T>),
-    Finalized(FinalizedRow<T>),
-}
-
-/// A finalized row, containing an indication which of the cells are known.
-/// Finalized rows cannot be changed and their range constraints are no longer available.
-#[derive(Clone)]
-pub struct FinalizedRow<T> {
-    /// The values, indices according to the columns stored in [FinalizableData].
-    values: Vec<T>,
-    /// A bit vector indicating which cells are known. Values of unknown cells should be ignored.
+struct CompactData<T: FieldElement> {
+    /// The ID of the first column used in the table.
+    first_column_id: u64,
+    /// The length of a row in the table.
+    column_count: usize,
+    /// The cell values, stored in row-major order.
+    data: Vec<T>,
+    /// Bit vector of known cells, stored in row-major order.
     known_cells: BitVec,
 }
 
-impl<T> FinalizedRow<T> {
-    pub fn new(values: Vec<T>, known_cells: BitVec) -> Self {
+impl<T: FieldElement> CompactData<T> {
+    /// Creates a new empty compact data storage. The column IDs have to be sorted.
+    fn new(column_ids: &[PolyID]) -> Self {
+        let first_column_id = column_ids.first().map_or(0, |id| id.id);
```

Comment:
maybe an assertion here that they are indeed sorted?

Reply:
the ids also need to be contiguous?

Reply:
No, they only need to be sorted.